ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessPackageUpload.scala

```diff
@@ -138,8 +108,53 @@ private[kvutils] case class ProcessPackageUpload(
     )
   }
 
+  private val preload: Runnable = () => {
+    val ctx = Metrics.preloadTimer.time()
+    try {
+      logger.trace("Preloading engine...")
+      val loadedPackages = engine.compiledPackages().packageIds
+      val packages: Map[Ref.PackageId, Ast.Package] = Metrics.decodeTimer.time { () =>
+        archives
+          .filterNot(
+            a =>
+              Ref.PackageId
+                .fromString(a.getHash)
+                .fold(_ => false, loadedPackages.contains))
+          .map { archive =>
+            Decode.readArchiveAndVersion(archive)._1
+          }
+          .toMap
+      }
+      packages.headOption.foreach {
+        case (pkgId, pkg) =>
+          engine
+            .preloadPackage(pkgId, pkg)
+            .consume(
+              _ => sys.error("Unexpected request to PCS in preloadPackage"),
+              pkgId => packages.get(pkgId),
+              _ => sys.error("Unexpected request to keys in preloadPackage")
+            )
+      }
+      logger.trace(s"Preload complete.")
+    } catch {
+      case scala.util.control.NonFatal(e) =>
+        logger.error("preload exception: $err")
+    } finally {
+      val _ = ctx.stop()
+    }
+  }
 }
 
-object ProcessPackageUpload {
-  val serialContext = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
+private[kvutils] object ProcessPackageUpload {
+  private[committing] val serialContext =
+    ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
+
+  private[committing] object Metrics {
+    private val registry = metrics.SharedMetricRegistries.getOrCreate("kvutils")
```

Comment:
Would it make sense to prefix this with something like the participant id?

Reply:
Not possible with current design where we process requests from many different participants and committers.

Reply:
Could you elaborate? How can the same kvutils instance (as in the jvm object) be used in multiple participants? We can call it something else, if the name `participantId` is the issue. Anyway, not blocking or major on my side. Just trying to understand :)

Reply:
This code is not used by participants. It is used by the committer to process a submission from a participant. We also use the same validator instance for multiple committers in development. But yes, it might make sense to include an identifier in the registry name for aggregation (though I guess that can also be solved at the point in which the metrics are collected).